server/src/main/java/io/crate/execution/engine/join/HashJoinBatchIterator.java

```diff
@@ -165,6 +198,36 @@ public boolean moveNext() {
         return true;
     }
 
+    private void fetchKeysFromBuffer() {
+        if (nonMatchingKeys == null) {
+            nonMatchingKeys = new ArrayList<>(buffer.size());
+            for (var values : buffer.entrySet()) {
+                if (values.getValue().v2() == false) {
+                    nonMatchingKeys.add(values.getKey());
+                }
+            }
+        }
+    }
+
+    private boolean emitNullValuesPairs() {
+        Integer key = nonMatchingKeys.getFirst();
+        List<Object[]> values = buffer.remove(key).v1();
+        Object[] nonMatches = values.removeFirst();
+
+        combiner.setLeft(new RowN(nonMatches));
+        combiner.setRight(new RowN(new Object[nonMatches.length]));
+
+        // This requires rehashing and we might be better off with
+        // storing the current nonMatches in an array
+        // and consume it from there
+        if (!values.isEmpty()) {
+            buffer.put(key, new Tuple<>(values, false));
```

Comment:
There are a few options here to to optimize: 

- Instead of reinserting to the buffer, we could just store all values in a list and consume it from there 
- If there is no limit a applied, we could fetch all values for all pairs and once and store them in a list, but this is inefficient when the batch iterator is not fully consumed.